### ISSUE TITLE
Bug: mypy incorrect error - does not recognize class as callable

### ISSUE THREAD

**Issue by vfrazao-ns1** [python/mypy#5361]
mypy incorrectly states that one of my objects is not callable when in fact it is.  Since the object is defined later in the file I am forced to use `from __future__ import annotations` to enter the type annotation.
Version info:
mypy 0.620 and Python 3.7
Error:
mypy error: `113: error: "Message" not callable`
Sample code (starting at line 113):
```Python
        msg = self.MSG_CLS[msg_type](*decoded_msg)
        return msg


@dataclass
class Message(object):
```

Message is indeed callable but mypy does not recognize that.

**Comment by ethanhs** [comment]
Hi, could you provide the source to this, or a minimal reproduction?

**Comment by ethanhs** [comment]
Ah, it looks like you are trying to instantiate a type, so your dict should be typed `Dict[int, Type[Message]]` not `Dict[int, Message]`. This will cause mypy to complain too many arguments are passed, which is correct I believe, since the base `Message` doesn't have any dataclass attributes, and uses `__slots__`.

**Comment by vfrazao-ns1** [comment]
Thank you. I'm brand new to mypy (and relatively new to programming). 

So far the project has been helpful - it's even caught a couple of mistakes for me.

**Comment by ethanhs** [comment]
No problem! Typing can take a little while to wrap your head around. Glad you've found mypy useful :)

**Comment by radioxoma** [comment]
Since 3.9 one could use `dict[int, type[Message]]`.

**Comment by yuwash** [comment]
In case someone came here from getting the error

```
app.py:9: error: "<typing special form>" not callable  [operator]
```

then it could be because you mistakingly assigned the `type[your_type]` to the name instead of annotating it.

Minimal example:

```python
class Pumpkin:
    pass


class Cook:
    food = type[Pumpkin]  # Wrong!
    # Above should’ve been `food: type[Pumpkin]`

    def serve(self) -> None:
        self.food()


class SimpleCook(Cook):
    food = Pumpkin


SimpleCook().serve()
```

Leaving the comment here as I didn’t find anything about “&lt;typing special form&gt;” anywhere.
